kv/kvserver: TestReplicaClosedTimestamp failed

kv/kvserver.TestReplicaClosedTimestamp [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8008755?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8008755?buildTab=artifacts#/) on master @ [93ed65565357538c9048ff45c878a493f2ed9b45](https://github.com/cockroachdb/cockroach/commits/93ed65565357538c9048ff45c878a493f2ed9b45):


```
=== RUN   TestReplicaClosedTimestamp
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp736341010
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestReplicaClosedTimestamp
    replica_closedts_internal_test.go:575: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp736341010
--- FAIL: TestReplicaClosedTimestamp (0.12s)
=== RUN   TestReplicaClosedTimestamp/sidetrans_closed_ahead
    replica_closedts_internal_test.go:572: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2167/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_closedts_internal_test.go:572
        	Error:      	Not equal: 
        	            	expected: hlc.Timestamp{WallTime:2, Logical:0, Synthetic:false}
        	            	actual  : hlc.Timestamp{WallTime:1, Logical:0, Synthetic:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (hlc.Timestamp) {
        	            	- WallTime: (int64) 2,
        	            	+ WallTime: (int64) 1,
        	            	  Logical: (int32) 0,
        	Test:       	TestReplicaClosedTimestamp/sidetrans_closed_ahead
    --- FAIL: TestReplicaClosedTimestamp/sidetrans_closed_ahead (0.04s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaClosedTimestamp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-22564

kv/kvserver.TestReplicaClosedTimestamp [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8287159?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8287159?buildTab=artifacts#/) on master @ [73485e56768d61b14a5aa9b3f880613d7820edb9](https://github.com/cockroachdb/cockroach/commits/73485e56768d61b14a5aa9b3f880613d7820edb9):


```
=== RUN   TestReplicaClosedTimestamp
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp16772011
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestReplicaClosedTimestamp
    replica_closedts_internal_test.go:575: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp16772011
--- FAIL: TestReplicaClosedTimestamp (0.15s)
=== RUN   TestReplicaClosedTimestamp/sidetrans_closed_ahead
    replica_closedts_internal_test.go:572: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3414/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_closedts_internal_test.go:572
        	Error:      	Not equal: 
        	            	expected: hlc.Timestamp{WallTime:2, Logical:0, Synthetic:false}
        	            	actual  : hlc.Timestamp{WallTime:1, Logical:0, Synthetic:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (hlc.Timestamp) {
        	            	- WallTime: (int64) 2,
        	            	+ WallTime: (int64) 1,
        	            	  Logical: (int32) 0,
        	Test:       	TestReplicaClosedTimestamp/sidetrans_closed_ahead
    --- FAIL: TestReplicaClosedTimestamp/sidetrans_closed_ahead (0.04s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaClosedTimestamp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestReplicaClosedTimestamp [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8456822?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8456822?buildTab=artifacts#/) on master @ [9caf75810ff7dbfa42d16b746c21c86bfa0c6324](https://github.com/cockroachdb/cockroach/commits/9caf75810ff7dbfa42d16b746c21c86bfa0c6324):


```
=== RUN   TestReplicaClosedTimestamp
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp2889241916
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestReplicaClosedTimestamp
    replica_closedts_internal_test.go:575: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp2889241916
--- FAIL: TestReplicaClosedTimestamp (0.13s)
=== RUN   TestReplicaClosedTimestamp/sidetrans_closed_ahead
    replica_closedts_internal_test.go:572: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3755/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_closedts_internal_test.go:572
        	Error:      	Not equal: 
        	            	expected: hlc.Timestamp{WallTime:2, Logical:0, Synthetic:false}
        	            	actual  : hlc.Timestamp{WallTime:1, Logical:0, Synthetic:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (hlc.Timestamp) {
        	            	- WallTime: (int64) 2,
        	            	+ WallTime: (int64) 1,
        	            	  Logical: (int32) 0,
        	Test:       	TestReplicaClosedTimestamp/sidetrans_closed_ahead
    --- FAIL: TestReplicaClosedTimestamp/sidetrans_closed_ahead (0.04s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaClosedTimestamp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestReplicaClosedTimestamp [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8552424?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8552424?buildTab=artifacts#/) on master @ [22244a780dcfaca48162dde8e0f90b5ba9b6bb9c](https://github.com/cockroachdb/cockroach/commits/22244a780dcfaca48162dde8e0f90b5ba9b6bb9c):


```
=== RUN   TestReplicaClosedTimestamp
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp277681229
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestReplicaClosedTimestamp
    replica_closedts_internal_test.go:575: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp277681229
--- FAIL: TestReplicaClosedTimestamp (0.09s)
=== RUN   TestReplicaClosedTimestamp/sidetrans_closed_ahead
    replica_closedts_internal_test.go:572: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3760/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_closedts_internal_test.go:572
        	Error:      	Not equal: 
        	            	expected: hlc.Timestamp{WallTime:2, Logical:0, Synthetic:false}
        	            	actual  : hlc.Timestamp{WallTime:1, Logical:0, Synthetic:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (hlc.Timestamp) {
        	            	- WallTime: (int64) 2,
        	            	+ WallTime: (int64) 1,
        	            	  Logical: (int32) 0,
        	Test:       	TestReplicaClosedTimestamp/sidetrans_closed_ahead
    --- FAIL: TestReplicaClosedTimestamp/sidetrans_closed_ahead (0.03s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaClosedTimestamp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestReplicaClosedTimestamp [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8568577?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8568577?buildTab=artifacts#/) on master @ [44d9f3c8b7bd46839187cea69eaec640c080ac05](https://github.com/cockroachdb/cockroach/commits/44d9f3c8b7bd46839187cea69eaec640c080ac05):


```
=== RUN   TestReplicaClosedTimestamp
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp4183852031
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestReplicaClosedTimestamp
    replica_closedts_internal_test.go:575: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestReplicaClosedTimestamp4183852031
--- FAIL: TestReplicaClosedTimestamp (0.10s)
=== RUN   TestReplicaClosedTimestamp/sidetrans_closed_ahead
    replica_closedts_internal_test.go:572: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/4372/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_closedts_internal_test.go:572
        	Error:      	Not equal: 
        	            	expected: hlc.Timestamp{WallTime:2, Logical:0, Synthetic:false}
        	            	actual  : hlc.Timestamp{WallTime:1, Logical:0, Synthetic:false}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,3 +1,3 @@
        	            	 (hlc.Timestamp) {
        	            	- WallTime: (int64) 2,
        	            	+ WallTime: (int64) 1,
        	            	  Logical: (int32) 0,
        	Test:       	TestReplicaClosedTimestamp/sidetrans_closed_ahead
    --- FAIL: TestReplicaClosedTimestamp/sidetrans_closed_ahead (0.03s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaClosedTimestamp.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_closedts_internal_test.go
+++ b/pkg/kv/kvserver/replica_closedts_internal_test.go
@@ -567,10 +567,13 @@ func TestReplicaClosedTimestamp(t *testing.T) {
 			cfg.ClosedTimestampReceiver = &r
 			tc.StartWithStoreConfig(ctx, t, stopper, cfg)
 			tc.repl.mu.Lock()
+			defer tc.repl.mu.Unlock()
 			tc.repl.mu.state.RaftClosedTimestamp = test.raftClosed
 			tc.repl.mu.state.LeaseAppliedIndex = uint64(test.applied)
-			tc.repl.mu.Unlock()
-			require.Equal(t, test.expClosed, tc.repl.GetCurrentClosedTimestamp(ctx))
+			// NB: don't release the mutex to make this test a bit more resilient to
+			// problems that could arise should something propose a command to this
+			// replica whose LeaseAppliedIndex we've mutated.
+			require.Equal(t, test.expClosed, tc.repl.getCurrentClosedTimestampLocked(ctx, hlc.Timestamp{} /* sufficient */))
 		})
 	}
 }

